### ISSUE TITLE
Bringing pull streams to modern JS

### ISSUE THREAD

**Issue by darsain** [pull-stream/pull-stream#74]
Pull streams are awesome, but their current implementation is limited to the JS from 10 years ago. This brings with it an awkward to use interface (callbacks everywhere), and limitations, some of which I described in #73.

This is a proposal to re-implement pull streams in modern JS.

## Source streams

Source (*readable*) streams are functions that create an async generator.

Example:

### values

```js
// Transforms an iterable object (such as array) into an async generator.
async function * values(iterable) {
	for (const item of iterable) {
		yield item;
	}
}
```

## Transform/through streams

Transform streams are functions that accept an async generator, and return another async generator.

Example streams:

### map

```js
// Applies `fn()` to all values before passing them through.
function map(fn) {
	return async function * (asyncGenerator) {
		for await (const value of asyncGenerator) {
			yield fn(value);
		}
	};
}
```

### filter

```js
// Lets through only values that pass the `test()`.
function filter(test) {
	return async function * (asyncGenerator) {
		for await (const value of asyncGenerator) {
			if (test(value)) {
				yield value;
			}
		}
	};
}
```

### delay

```js
// Delays each read by `ms`.
function delay(ms) {
	return async function * (asyncGenerator) {
		for await (const value of asyncGenerator) {
			await new Promise(resolve => setTimeout(resolve, ms));
			yield value;
		}
	};
}
```

### errorOn

```js
// Throws an error when value passing through matches `x`.
function errorOn(x) {
	return async function * (asyncGenerator) {
		for await (const value of asyncGenerator) {
			if (value === x) {
				throw new Error(`Value can't be "${x}"`);
			}
			yield value;
		}
	};
}
```

## Sink (*writable*) streams

Sinks are functions that accept an async generator, and return a promise.

Example sinks:

### log

```js
// Logs all values pulled from passed stream.
async function log(asyncGenerator) {
	for await (const value of asyncGenerator) {
		console.log(value);
	}
}
```

### reduce

```js
// Reduces all values with reducer().
function reduce(reducer, acc) {
	return async function (asyncGenerator) {
		for await (const value of asyncGenerator) {
			acc = reducer(acc, value);
		}
		return acc;
	}
}
```

# Examples

Assuming this little pull helper:

```js
// Help us compose streams left to right.
function pull(...args) {
	let result = args.reverse().shift();
	for (let stream of args) {
		result = result(stream);
	}
	return result;
}
```

Read numbers 0 to 4 and sum them up:

```js
const arr = Array(5).fill(0).map((x, i) => i); // array [0..4]
const numbers = values(arr); // source stream of numbers from arr
const sum = reduce((x, y) => x + y, 0); // sink to sum up all values

// raw
sum(numbers).then(x => console.log(x));

// with pull helper
pull(numbers, sum).then(x => console.log(x));

// in an async function
(async () => {
	console.log(await pull(numbers, sum));
})();

// All versions above log:
// > 10
```

Read numbers 0 to 4, delay each read by 300ms, filter only even numbers, and square the rest:

```js
const arr = Array(5).fill(0).map((x, i) => i); // array [0..4]
const numbers = values(arr); // source stream of numbers from arr
const wait300 = delay(300); // transform stream to add 300 ms delay to each read
const even = filter(x => x % 2 === 0); // transform stream to filter only even numbers
const square = map(x => x * x); // transform stream to square each number

// raw
log(square(even(wait300(numbers)))).then(() => console.log('done'));

// with pull helper
pull(numbers, wait300, even, square, log).then(() => console.log('done'));

// in an async function
(async () => {
	await pull(numbers, wait300, even, square, log);
	console.log('done');
})();

// All versions above log:
// > 0
// > 4
// > 16
// > done
```

Read numbers 0 to 4, but throw an error on 2:

```js
const arr = Array(5).fill(0).map((x, i) => i);

pull(values(arr), errorOn(2), log).then(null, err => console.log(`Error: ${err.message}`));

// in an async function
(async () => {
	try {
		await pull(values(arr), errorOn(3), log);
	} catch(err) {
		console.error(`Error: ${err.message}`);
	}
})();

// All versions above log:
// > 0
// > 1
// > Error: Value can't be "2"
```

# Comparison with current implementation

The implementation above has it all. Backpressure, error propagation, and more:

### Simpler to work with

You don't need to be passing callbacks into sinks and doing other stuff from 10 years ago.

Do you want the final value from reducer? Just `await`, or `.then()` it to something else.

Do you want to catch errors? `try/catch` or add `.then(null, onError)` handler.

It all just seamlessly fits into modern async control flow.

### Easily identifiable interfaces

Sources and transforms are async generators, while sinks are promises. We can duck type and work with all of these interfaces already. No flags required.

### Simpler to implement

Everything looks like sync code. No callbacks juggling whatsoever. Non-blocking by default due to nature of promises. No fear of blowing some stinky call stack. Just look how simple each stream constructor looks like!

### Easier to understand and reason about

Once you understand async functions & async generators, the brain power required to thinking about, use, and implement these async interfaces drops to an absolute minimum. No callbacks in sight.

# Environment support

`async/await` is gonna land in node behind a flag October 2016 (this month), and without a flag possibly mid December (already in Chrome Canary).

Async generators is currently stage 3 (candidate), which means the spec is complete, and browser implementations are going to be landing any moment.

Regardless, you can use all of it already with babel transpilation.

---

*And not a single callback was passed that day.*

**Comment by dominictarr** [comment]
those are the easy ones. show me how to implement

https://github.com/pull-stream/pull-cat (just to warm up)
https://github.com/pull-stream/pull-merge
https://github.com/pull-stream/pull-many

then we'll compare perf and simplicity.
one thing that concerns me about await is that is gives you this faux-blocking systax, which makes it _so easy_ to write serial code - that leads to people writing serial code that waits unnecessarily...

(for example - cargo, the rust package manager, does this even though rust is meant to be all about modern async!)

**Comment by dominictarr** [comment]
also, adding promises is a very _seperate_ argument to async. that is a breaking change, you should hold off on that for now, remembering there are a lot more pull-streams than just this library

**Comment by darsain** [comment]
> then we'll compare simplicity

Callbacks can never beat modern async flow control in simplicity :)

> and perf

The current implementation is definitely faster than my proposal, especially since it has to be transpiled into quite monstrous code 😄. But as these features land, and engines optimize them away, the perf will become very similar. Plus, in most of the real world streams (at least in gulp), most of the time is spent by processing data by tools, not by the streaming interface itself.

> one thing that concerns me about await is that is gives you this faux-blocking systax, which makes it so easy to write serial code

If you don't see the benefits, and are against using async/await and the rest of the modern JS features in general, than I guess there is nothing more for me to do here.

> that leads to people writing serial code that waits unnecessarily...

You can't stop people from shooting themselves in the foot. And it's definitely easier to accomplish that in the world of passing callbacks everywhere.

> those are the easy ones. show me how to implement...

Oh boy, lets go :)

### cat

Trivial:

```js
async function * cat(generators) {
	for (const generator of generators) {
		for await (const value of generator) {
			yield value;
		}
	}
}
```

### many

Trivial, maybe could be simplified somehow:

```js
async function * many(generators) {
	generators = generators.slice();
	let i = 0;

	while (generators.length) {
		const {value, done} = await generators[i].next();

		if (done) {
			generators.splice(i--, 1);
		} else {
			yield value;
		}

		i = ++i % generators.length;
	}
}
```

### merge

This one was interesting and took me a bit to implement. I'm definitely not happy with it. I'm sure it can be done in a cleaner way, especially without the need for that ugly `preYields` argument, but It's almost midnight in here, I want to go watch some movie and hit the bed, so this will have to do for now :) And hey, I still didn't even need recursion! 😄

Also, it *might* have bugs, I've run it only through a few test examples I copy & pasted from your pull-merge tests folder, but nothing comprehensive.

```js
function defaultCompare(a, b) {
	return a < b ? -1 : a > b ? 1 : 0;
}

async function * merge(left, right, compare = defaultCompare, preYields = 0) {
	if (Array.isArray(left)) {
		let streams = left.slice();
		left = streams.shift();

		while (streams.length >= 2) {
			left = merge(left, streams.shift(), compare, preYields++);
		}

		right = streams.shift();
	}

	let rightDone;

	leftLoop:
	for await (let leftValue of left) {
		for (let i = preYields; i--;) {
			yield leftValue;
			let obj = await left.next();
			if (obj.done) {
				break leftLoop;
			}
			leftValue = obj.value;
		}

		if (rightDone) {
			yield leftValue;
			continue;
		}

		const {value: rightValue, done} = await right.next();

		if (done) {
			yield leftValue;
			rightDone = true;
			continue;
		}

		let comparison = compare(leftValue, rightValue);

		if (comparison === 0) {
			yield rightValue;
		} else {
			yield comparison < 0 ? leftValue : rightValue;
			yield comparison < 0 ? rightValue : leftValue;
		}
	}

	for await (const value of right) {
		yield value;
	}
}
```

**Comment by dominictarr** [comment]
@darsain I'm not sure that your _many_ does the same thing as mine, could you provide me with runnable code? I tried running this with node and got syntax errors. I'd like to actually measure performance, and check whether the behavior is actually correct.

> The current implementation is definitely faster than my proposal, especially since it has to be transpiled into quite monstrous code :smile:. But as these features land, and engines optimize them away, the perf will become very similar. Plus, in most of the real world streams (at least in gulp), most of the time is spent by processing data by tools, not by the streaming interface itself.

in the case of gulp I agree that performance is not everything, but this is also used other cases where high performance means we can use pull streams for very light weight tasks - or things that can get very hot like parsers.

Given the modules inside of pull-stream are already correct, and are basically perfect, I don't see a need to rewrite them...

However, I don't want to stop you from writing your pull-streams with babel if that is how you wish to do it. Is there anything about pull-stream that prevents mixing callback pull-streams with generator pull-streams?

**Comment by nichoth** [comment]
Since it is a module that is meant to be consumed (a dependency) I appreciate that I don't have to set up or think about transpilers in order to use it. Since moving to new JS syntax would break the entire interface, I don't see any reason why a modern JS implementation shouldn't be a different package.

I think the underlying issue here is that it is difficult to create a correct source or sink, which is true, but it should be ok because there should be an existing stream operator for any use case.

**Comment by darsain** [comment]
Yeah, I see that creating a separate project would be the best course of action here. I'll get on it soon.

@dominictarr here is a minimal `package.json` to test this stuff locally:

```js
{
  "devDependencies": {
    "babel-cli": "^6.16.0",
    "babel-preset-stage-3": "^6.17.0"
  },
  "scripts": {
    "start": "babel-node index.js"
  },
  "babel": {
    "presets": ["stage-3"]
  }
}
```

Just put it somewhere, run `npm isntall`, write your stuff into `index.js`, and run it with `npm start`. `babel-node` command comes with `babel-cli` package.

**Comment by ahdinosaur** [comment]
hey @darsain, this is really cool to see an alternative pull stream implementation using promises, generators, iterators, and async / await. keen to see it published as a module, it's seriously good! however i personally stay away from the above features in "modern JS", because i _prefer_ "luddite JS" using callbacks, since that's my opinion.

**Comment by TehShrike** [comment]
I would be interested in using the newer features, myself, but I think it's reasonable to keep the two implementations separate, since I'm sure there will be plenty of people who prefer callback APIs even once the newer features become standard.

I don't know if @darsain has the time to start the project himself - I don't have the development bandwidth to be a major contributor at the moment, but I would definitely start lurking.  But I think a project that aimed for feature parity with the basic pull-stream set using async/await/promises would be excellent to have.  That's the version I would probably use first in a new project.

Perhaps a link in the pull-stream docs, something like "for a version of pull-stream using async/await and promises, click here" could be added eventually in case people come looking for pull streams who prefer the newer features when available.

**Comment by darsain** [comment]
> I don't know if @darsain has the time to start the project himself

Already started on it :)

**Comment by dominictarr** [comment]
@darsain i think it would be easier for everyone if you just posted your code in a way we could clone and run npm install. Use a secret gist if you don't wanna have it on your github yet.

**Comment by brucou** [comment]
This certainly helps clarify some conceptual parts. However I am not sure I see the value here:
- proposed implementation is slower, and is likely to remain so in the foreseeable future (async/yield syntax is state machine + callback, so cant be faster than directly using callbacks)
- requires intermediate steps (babel transpilation)
- is more complicated to install/build/etc.

So what's the REAL benefits here, apart from this 'modern' hype for a end-user ?? And I am not all against ES6-7-X. But first a end-user programs against an API, so implementation details are irrelevant to him (till you have to debug, that is - which is where the intermediate steps complicate things).

**Comment by TehShrike** [comment]
> proposed implementation is slower, and is likely to remain so in the foreseeable future (async/yield syntax is state machine + callback, so cant be faster than directly using callbacks)

For some developers, the code is more important than the difference in overhead between functions and async/await.

> requires intermediate steps (babel transpilation)

For many developers, this is already a fact of life.

> is more complicated to install/build/etc.

I don't believe this is accurate.  My workflow is `npm i library -S` + `require` either way.

@darsain's library would be the [Koa](http://koajs.com/) to pull-stream's [express](http://expressjs.com/).  Just because the original exists doesn't mean there won't be a good audience for the one using newer language features.

Do you have a link to the repo @darsain?

**Comment by darsain** [comment]
> proposed implementation is slower

Yeah, temporarily. But maybe not even by much. I need to get some benchmarks done, but first the babel issue linked below needs to get fixed.

> is more complicated to install/build/etc
> requires intermediate steps (babel transpilation)

Nope. Install is same as any other module, since the thing published to npm is already transpiled. In fact, if you don't need to write any streams yourself (most of the time you'll get by just with `map` & `filter`), you won't even need the transpilation step. And if you do want to implement your own streams, well babel is pretty much standard nowadays. And for a good reason. Working in ES5- is just plain painful compared to modern ES.

To sum up *again*: Async iterator implementation provides simpler code which is easier to reason about, implement, work with, and will be even easier to debug. As @TehShrike mentioned, you'd choose this kind of implementation for the same reason you'd choose koa over express.

---

@TehShrike I have a lot of work done locally, but can't move forward as https://github.com/babel/babel/issues/4783 needs to to get fixed first before I can finish it all up, transpile, write tests, and publish to npm. That's already 2nd blocking issue from babel I encountered working on this, hopefully it'll be the last.

I would publish to github, but haven't decided on the name yet. I was expecting it to be **pul**, which was so simple and fitting, I even had a nice logo done, but while I was waiting for the babel patch someone else snatched it on npm 😢. So now I'm looking for an alternative. Open to suggestions 😃.

**Comment by TehShrike** [comment]
`apull` is a cute homophone.

**Comment by m-onz** [comment]
Sorry to barge in but I agree whole heartily with @ahdinosaur pull streams are perfect as they are and allow us to make light weight constructs on top of it. I don't want to have to be forced to use babel at the lowest level and go through pains to support node v0.12 when making packages. I would love to see some npm packages that are stream based around generators or iterators but it seems silly to change the pull stream core when its easy to convert a generator to a stream and via versa. Basically please don't change pull streams to use the new modern JS... personally I don't like the async and wait stuff.

**Comment by darsain** [comment]
@m-onz yeah, agreed. We pretty much established that :) The new packages are almost done, but need some async iterator babel bugs to get fixed before they can be finished and published.

**Comment by m-onz** [comment]
@darsain Yes sorry to re-enforce the same opinion again but I feel strongly about it! haha... That being said I would love to check out your package when its ready as it does look really interesting.. it reminded me of this: https://github.com/casperin/generator-pull-streams and this: https://gist.github.com/pmuellr/90b852cac2d63f18e15773097e0340d4 btw? as they look pretty similar to what your working on.

**Comment by taoeffect** [comment]
I really like this idea.

Are there any implementations? Also, would it make sense to combine this in some way with SICP style streams ala [stream.js](https://dionyziz.github.io/stream.js-website/)?

**Comment by harrysarson** [comment]
@darsain Did you make any progress with this?

**Comment by darsain** [comment]
@harrysarson back when I needed this kind of streams I couldn't finish it because of a blocking babel transpilation issue, and by the time they fixed it I no longer needed it and moved on to other things. The project is still idling on my disk. I hope to sometime return to it, but I have tons of higher priority things on my todo list atm :(

**Comment by harrysarson** [comment]
@darsain fair enough :)
